pkg/provider/consulcatalog/consul_catalog.go

```diff
@@ -294,6 +347,136 @@ func contains(values []string, val string) bool {
+		if raw == nil {
+			return
+		}
+
+		v, ok := raw.(*api.CARootList)
+		if !ok || v == nil {
+			log.FromContext(ctx).Errorf("invalid result for root certificate watcher")
+			return
+		}
+
+		roots := make([]string, 0, len(v.Roots))
+		for _, root := range v.Roots {
+			roots = append(roots, root.RootCertPEM)
+		}
+
+		dest <- roots
+	}
+}
+
+type keyPair struct {
+	cert string
+	key  string
+}
+
+func leafWatcherHandler(ctx context.Context, dest chan<- keyPair) func(watch.BlockingParamVal, interface{}) {
+	return func(_ watch.BlockingParamVal, raw interface{}) {
+		if raw == nil {
+			return
+		}
+
+		v, ok := raw.(*api.LeafCert)
+		if !ok || v == nil {
+			log.FromContext(ctx).Errorf("invalid result for leaf certificate watcher")
+			return
+		}
+
+		dest <- keyPair{
+			cert: v.CertPEM,
+			key:  v.PrivateKeyPEM,
+		}
+	}
+}
+
+func (p *Provider) watchConnectTLS(ctx context.Context) {
+	ctxLog := log.With(ctx, log.Str(log.ProviderName, "consulcatalog"))
+	logger := log.FromContext(ctxLog)
+
+	leafWatcher, err := watch.Parse(map[string]interface{}{
+		"type":    "connect_leaf",
+		"service": p.ServiceName,
+	})
+	if err != nil {
+		logger.WithError(err).Error("failed to create leaf cert watcher plan")
+		return
+	}
+
+	rootWatcher, err := watch.Parse(map[string]interface{}{
+		"type": "connect_roots",
+	})
+	if err != nil {
+		logger.WithError(err).Error("failed to create root cert watcher plan")
+	}
+
+	leafChan := make(chan keyPair)
+	rootChan := make(chan []string)
+
+	leafWatcher.HybridHandler = leafWatcherHandler(ctx, leafChan)
+	rootWatcher.HybridHandler = rootsWatchHandler(ctx, rootChan)
+
+	logOpts := &hclog.LoggerOptions{
+		Name:       "consulcatalog",
+		Level:      hclog.LevelFromString(logrus.GetLevel().String()),
+		JSONFormat: true,
+	}
+
+	hclogger := hclog.New(logOpts)
+
+	go func() {
+		err := leafWatcher.RunWithClientAndHclog(p.client, hclogger)
+		if err != nil {
+			logger.WithError(err).Errorf("Leaf certificate watcher failed with error")
+		}
+	}()
+
+	go func() {
+		err := rootWatcher.RunWithClientAndHclog(p.client, hclogger)
+		if err != nil {
+			logger.WithError(err).Errorf("Root certificate watcher failed with error")
+		}
+	}()
+
+	leafCerts := <-leafChan
+	rootCerts := <-rootChan
+
+	certInfo := &connectCert{
+		service: p.ServiceName,
+		root:    rootCerts,
+		leaf:    leafCerts,
+	}
+
+	logger.Debugf("Received connect certs for service %s", p.ServiceName)
+	p.certChan <- certInfo
+
+	ticker := time.NewTicker(time.Duration(p.RefreshInterval))
+
+	for {
+		select {
+		case <-ctx.Done():
+			ticker.Stop()
+			return
+
+		case rootCerts = <-rootChan:
+		case leafCerts = <-leafChan:
+
+		case <-ticker.C:
+			newCertInfo := &connectCert{
+				service: p.ServiceName,
+				root:    rootCerts,
+				leaf:    leafCerts,
+			}
+			if !reflect.DeepEqual(newCertInfo, certInfo) {
+				logger.Debugf("Updating connect certs for service %s", p.ServiceName)
+				p.certChan <- newCertInfo
+			}
+		}
```

Comment:
Do we actually need/want a ticker here? Could we simply remove the ticker and do all that if either rootCerts or leafCerts is refreshed (i.e. after the select block, just like you did in the operation closure in Provide) ?

Reply:
We do not "need" the ticker, but it would prevent multiple certs being sent down in quick succession. But I guess for simplicity reasons we can just drop the ticker.